File: web/src/content.ts
/**
 * What a restricted face may show.
 *
 * A proxy in front of the server can give one hostname to the music and
 * another to the videos (`X-Media-Content`), and the server answers
 * accordingly whatever the page asks. This is only the page's side of it:
 * which chips are worth offering, and which view to fall back to when a
 * link arrives for one that is not there.
 */

/** The views the page has, as they appear in the URL hash. */
export type ViewMode =
  | 'all'
  | 'video'
  | 'image'
  | 'audio'
  | 'started'
  | 'watched'
  | 'popular'
  | 'albums'
  | 'artists'
  | 'genres'
  | 'audiobooks'
  | 'series';

/**
 * Whether a view is worth offering.
 *
 * With nothing withheld, all of them are. With one class of media, the chip
 * for that class would only repeat "All" — the whole library *is* that class
 * — so it is left out, and what remains is All plus whatever grouped views
 * the class has: albums and artists for music, nothing of the sort for
 * pictures or films. Restricted to two classes the chips come back, because
 * now they choose between something.
 */
export function modeShown(mode: ViewMode, content: string[] | null): boolean {
  if (!content || content.length === 0) return true;
  const has = (c: string): boolean => content.includes(c);
  switch (mode) {
    case 'all':
      return true;
    case 'albums':
    case 'artists':
    case 'genres':
    case 'audiobooks':
      return has('music');
    case 'video':
      return has('videos') && content.length > 1;
    case 'series':
      // Television is video, so this belongs wherever films do — and it
      // says something All does not, so it stays even when films are all
      // there is.
      return has('videos');
    case 'popular':
      // Anything can be played, so this belongs on every face — unlike the
      // two watch views, which list films.
      return true;
    case 'started':
    case 'watched':
      // These list films, so they belong wherever films do — and unlike the
      // Videos chip they say something All does not, so they stay even when
      // films are all there is.
      return has('videos');
    case 'image':
      return has('images') && content.length > 1;
    case 'audio':
      return has('music') && content.length > 1;
  }
}

/**
 * Where a face opens when the URL says nothing.
 *
 * A library that is only music is a library of *performers*: the file
 * listing there is a thousand tracks in whatever order they were written to
 * disk, which is the least useful thing to be shown first. Everywhere else
 * the listing is right — a mixed library has no one shelf to open on, and a
 * face of films has no grouping to open into.
 *
 * Only the default: a link that names a view still opens that view, and the
 * chips are unchanged.
 */
export function defaultMode(content: string[] | null): ViewMode {
  return content?.length === 1 && content[0] === 'music' ? 'artists' : 'all';
}

/** The view to open when the one asked for is not on offer here. */
export function fallbackMode(mode: ViewMode, content: string[] | null): ViewMode {
  return modeShown(mode, content) ? mode : 'all';
}

/** What a view's "queue all" asks the server for: how its tracks are gathered. */
export type QueueSource = 'albums' | 'artists' | 'genres' | 'audiobooks' | 'items';

/**
 * Where "queue all" is offered, and what it queues.
 *
 * The three grouped views of music queue themselves — every release listed,
 * every release of every performer listed, every release in every genre
 * listed. The music chip queues its tracks. The listings that mix kinds
 * queue only on a face that shows nothing but music, where they are track
 * listings too; elsewhere a button over a grid of films that queued the odd
 * track among them would be a puzzle, not a control.
 */
export function queueSource(mode: ViewMode, content: string[] | null): QueueSource | null {
  switch (mode) {
    case 'albums':
    case 'artists':
    case 'genres':
    case 'audiobooks':
      return mode;
    case 'audio':
      return 'items';
    case 'all':
    case 'popular':
      return content?.length === 1 && content[0] === 'music' ? 'items' : null;
    default:
      return null;
  }
}

File: internal/server/content.go
package server

// One library, several faces.
//
// The same server can answer as a music library to one caller and as a video
// library to another: a request carrying `X-Media-Content: music` is shown
// music and nothing else — not in its listings, not in its counts, and not
// if it asks for a video by id.
//
// The header is set by whatever sits in front (a reverse proxy giving one
// hostname to each face), never by the page: a browser cannot put headers on
// the requests that matter here — an <img> asking for a thumbnail or a
// <video> asking for a stream send what they like — which is exactly why the
// filtering is done here rather than in the client. The client is told what
// it may see (`/api/info`) so that it can leave out the views it has no
// business offering, but that is a courtesy; this is the enforcement.
//
// Indexing is untouched. The library holds everything it always did, one
// scan feeds every face, and turning the header off shows the whole library
// again with nothing to rebuild.

import (
	"net/http"
	"strings"

	"github.com/JohanLindvall/Mediator/internal/library"
)

// ContentHeader carries the classes of media a request may see, comma
// separated: "music", "videos", "images". Absent or empty means everything.
const ContentHeader = "X-Media-Content"

// content is the set of classes one request is allowed to see. Music covers
// both audio files and the playlists that list them: a playlist is how a
// release names its own running order, and a music library without them
// would be missing half of what it groups by.
type content struct {
	video bool
	image bool
	music bool
}

var everything = content{video: true, image: true, music: true}

// contentOf reads the restriction a request arrived with.
func contentOf(r *http.Request) content { return parseContent(r.Header.Get(ContentHeader)) }

// parseContent reads the header. Anything it does not recognise is ignored;
// a header naming nothing recognisable is treated as no header at all,
// because the alternative is a face that shows nothing and looks broken.
func parseContent(h string) content {
	var c content
	for _, part := range strings.Split(h, ",") {
		switch strings.ToLower(strings.TrimSpace(part)) {
		case "music", "audio":
			c.music = true
		case "video", "videos":
			c.video = true
		case "image", "images", "photos", "pictures":
			c.image = true
		}
	}
	if c == (content{}) {
		return everything
	}
	return c
}

// unrestricted reports whether this is the whole library.
func (c content) unrestricted() bool { return c == everything }

// allows reports whether an item of this kind may be seen at all.
func (c content) allows(k library.Kind) bool {
	switch k {
	case library.KindVideo:
		return c.video
	case library.KindImage:
		return c.image
	case library.KindAudio, library.KindPlaylist:
		return c.music
	}
	return false
}

// The sets the three faces draw from, built once: kinds is asked several
// times per request.
var (
	videoKinds = library.KindsOf(library.KindVideo)
	imageKinds = library.KindsOf(library.KindImage)
	musicKinds = library.KindsOf(library.KindAudio, library.KindPlaylist)
)

// kinds is what a listing may draw from.
func (c content) kinds() library.KindSet {
	if c.unrestricted() {
		return 0 // the zero set is every kind
	}
	var set library.KindSet
	if c.video {
		set |= videoKinds
	}
	if c.image {
		set |= imageKinds
	}
	if c.music {
		set |= musicKinds
	}
	return set
}

// mask restricts totals to what may be seen. The per-kind numbers are
// independent, so this is subtraction rather than a second count — and the
// total is recomputed from what is left rather than carried over, or a face
// showing music would report the whole library's size beside it.
func (c content) mask(x library.Counts) library.Counts {
	if c.unrestricted() {
		return x
	}
	if !c.video {
		x.Video = 0
	}
	if !c.image {
		x.Image = 0
	}
	if !c.music {
		x.Audio, x.Playlist, x.Albums, x.Artists, x.Genres, x.Audiobooks = 0, 0, 0, 0, 0, 0
	}
	if !c.video {
		x.Series = 0 // television is video, and this face has none
	}
	x.Total = x.Video + x.Image + x.Audio + x.Playlist
	return x
}

// names is what /api/info tells the client it may show, in a fixed order so
// the answer does not change shape between requests.
func (c content) names() []string {
	if c.unrestricted() {
		return nil // nothing withheld: the client offers everything
	}
	var out []string
	if c.video {
		out = append(out, "videos")
	}
	if c.image {
		out = append(out, "images")
	}
	if c.music {
		out = append(out, "music")
	}
	return out
}

// item resolves the item a request names, and refuses one this request is
// not allowed to see — a direct link to a video is nothing on a music face.
func (s *Server) item(r *http.Request, id string) (library.Item, bool) {
	it, ok := s.lib.Get(id)
	if !ok || !contentOf(r).allows(it.Kind) {
		return library.Item{}, false
	}
	// And the other restriction, which composes with it: whichever of the
	// two says no, the answer is that this caller has no such item. Every
	// by-id handler resolves through here — the stream, the thumbnail, the
	// conversions, the flags, the positions, the subtitles — so this is what
	// makes a path restriction hold for all of them at once.
	if !pathsOf(r).AllowsItem(it) {
		return library.Item{}, false
	}
	return it, true
}

// PathsHeader restricts a request to the parts of the library that live
// under the directories it names, separated by commas (or newlines, for a
// path with a comma in it). Absent or empty means the whole library.
//
// The same posture as the content header above and set the same way — by
// whatever sits in front, never by the page — and for the same reason: a
// browser cannot put a header on the <img> that fetches a thumbnail or the
// <video> that fetches a stream, so the filtering has to be here.
//
// The two compose. A request carrying both is shown the intersection, which
// falls out of each being applied where it belongs rather than either
// knowing anything about the other.
const PathsHeader = "X-Allowed-Paths"

// pathsOf reads the restriction a request arrived with.
func pathsOf(r *http.Request) library.PathFilter {
	return library.ParsePaths(r.Header.Get(PathsHeader))
}
